typescript/src/deposit-sweep.ts

```diff
@@ -130,238 +150,297 @@ export async function assembleDepositSweepTransaction(
       outputIndex: 0, // There is only one output.
-      value: BigNumber.from(transaction.outputs[0].value),
+      value: BigNumber.from(transaction.outs[0].value),
     },
     rawTransaction: {
-      transactionHex: transaction.toRaw().toString("hex"),
+      transactionHex: transaction.toHex(),
     },
   }
 }
 
 /**
- * Creates script for the transaction input at the given index and signs the
- * input.
- * @param transaction - Mutable transaction containing the input to be signed.
- * @param inputIndex - Index that points to the input to be signed.
- * @param walletKeyRing - Key ring created using the wallet's private key.
- * @returns Empty promise.
+ * Signs the main UTXO transaction input and sets the appropriate script or
+ * witness data.
+ * @param transaction - The transaction containing the input to be signed.
+ * @param inputIndex - Index pointing to the input within the transaction.
+ * @param prevOutScript - The previous output script for the input.
+ * @param prevOutValue - The value from the previous transaction output.
+ * @param keyPair - A Signer object with the public and private key pair.
+ * @param network - The Bitcoin network type (mainnet or testnet).
+ * @returns An empty promise upon successful signing.
+ * @throws Error if the UTXO doesn't belong to the wallet, or if the script
+ *         format is invalid or unknown.
  */
 async function signMainUtxoInput(
-  transaction: any,
+  transaction: Transaction,
   inputIndex: number,
-  walletKeyRing: any
+  prevOutScript: Buffer,
+  prevOutValue: number,
+  keyPair: Signer,
+  network: networks.Network
 ) {
-  const previousOutpoint = transaction.inputs[inputIndex].prevout
-  const previousOutput = transaction.view.getOutput(previousOutpoint)
-  if (!walletKeyRing.ownOutput(previousOutput)) {
+  if (!ownsUtxo(keyPair, prevOutScript, network)) {
     throw new Error("UTXO does not belong to the wallet")
   }
-  // Build script and set it as input's witness
-  transaction.scriptInput(inputIndex, previousOutput, walletKeyRing)
-  // Build signature and add it in front of script in input's witness
-  transaction.signInput(inputIndex, previousOutput, walletKeyRing)
+
+  const sigHashType = Transaction.SIGHASH_ALL
+
+  if (isP2PKH(prevOutScript)) {
+    // P2PKH
+    const sigHash = transaction.hashForSignature(
+      inputIndex,
+      prevOutScript,
+      sigHashType
+    )
+
+    const signature = script.signature.encode(
+      keyPair.sign(sigHash),
+      sigHashType
+    )
+
+    const scriptSig = payments.p2pkh({
+      signature: signature,
+      pubkey: keyPair.publicKey,
+    }).input!
+
+    transaction.ins[inputIndex].script = scriptSig
+  } else if (isP2WPKH(prevOutScript)) {
+    // P2WPKH
+    const decompiledScript = script.decompile(prevOutScript)
+    if (
+      !decompiledScript ||
+      decompiledScript.length !== 2 ||
+      decompiledScript[0] !== 0x00 ||
+      !Buffer.isBuffer(decompiledScript[1]) ||
+      decompiledScript[1].length !== 20
```

Comment:
Can you post a comment explaining this black magic? 😅

Reply:
In 3fa34d91cf1ded905b7f5abf9691f91144df09cb I managed to significantly simplify this code.
The `if` was removed.